### PR TITLE
Functions for more precise time measurement

### PR DESCRIPTION
**Pull request by pali6**
Adds `time_reset`, `time_microseconds` and `time_milliseconds` BYOND-accessible functions which can be used to create and reset multiple "timers" and track time since they were last reset more precisely than it is possible in BYOND itself. I personally mostly intend to use these for some detailed live profiling in some performance-problematic parts of https://github.com/goonstation/goonstation but maybe other people can use these too.

I tested that these work but yell at me if I forgot anything or screwed up as Rust is not the language I use most of the time, thanks.

### CODE REVIEW

File: src/time.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+use std::{
+    cell::RefCell,
+    collections::hash_map::{Entry, HashMap},
+    ffi::OsString,
+    time::Instant,
+};
+
+thread_local!( static INSTANTS: RefCell<HashMap<OsString, Instant>> = RefCell::new(HashMap::new()) );
```
Comment (SpaceManiac) [inline src/time.rs:8]: Using OsString here is likely unnecessary; BYOND has also standardized on UTF-8, just like Rust Strings.

**Comment by SpaceManiac** [comment]
With regard to profiling, [be aware](https://doc.rust-lang.org/std/time/struct.Instant.html):
> Note, however, that instants are not guaranteed to be steady. In other words, each tick of the underlying clock might not be the same length (e.g. some seconds may be longer than others). An instant may jump forwards or experience time dilation (slow down or speed up), but it will never go backwards.

However, the effect is likely to be small.

File: src/time.rs (hunk 1 above)
Comment (pali6) [reply src/time.rs:8]: Ah, honestly I just loked at whatever `log.rs` used I think because I wasn't sure.

**Comment by pali6** [comment]
Hm, I'm personally fine with accepting the risk of unsteady time and dunno what else to use. Does `time::SteadyTime` have comparable precision and would it be a better fit?

**Comment by SpaceManiac** [comment]
Considering the `time` crate removed `SteadyTime` in v0.3, they must not have liked it very much, but I don't have any details for or against it myself.

**Comment by pali6** [comment]
Oh lol, I must have had an old version of the docs opened up.

**Comment by ZeWaka** [comment]
@tgstation/rust-g-dev-team could we get this reviewed?

File: src/lib.rs [hunk 2]
```diff
@@ -34,6 +34,8 @@ pub mod unzip;
 pub mod url;
 #[cfg(feature = "worleynoise")]
 pub mod worleynoise;
+#[cfg(feature = "time")]
+pub mod time;
```
Comment (Mothblocks) [inline src/lib.rs:38]: i can't auto suggest it, but put this in alphabetic order like the rest of them

File: Cargo.toml [hunk 3]
```diff
@@ -58,6 +58,7 @@ log = ["chrono"]
 sql = ["mysql", "serde", "serde_json", "once_cell", "dashmap", "jobs"]
 toml = ["serde", "serde_json", "toml-dep"]
 url = ["url-dep", "percent-encoding"]
+time = []
```
Comment (Mothblocks) [inline Cargo.toml:61]: abc order

File: dmsrc/time.dm [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#define rustg_time_microseconds(id) call(RUST_G, "time_microseconds")(id)
+#define rustg_time_milliseconds(id) call(RUST_G, "time_milliseconds")(id)
```
Comment (Mothblocks) [inline dmsrc/time.dm:2]: You should maybe have these call `text2num` if they're going to return as strings

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by AffectedArc07** [comment]
>  I personally mostly intend to use these for some detailed live profiling in some performance-problematic parts of https://github.com/goonstation/goonstation 

This is just btime 2021 :P

**Comment by pali6** [comment]
@tgstation/rust-g-dev-team are there any further issues with this PR?

**Comment by optimumtact** [comment]
@tgstation/rust-g-dev-team 
Will merge pending no further objections in 24 hours.

**Pull request merged**